client/src/main/java/io/pravega/client/netty/impl/ConnectionPoolImpl.java

```diff
@@ -0,0 +1,279 @@
+        Exceptions.checkNotClosed(closed.get(), this);
+
+        final List<Connection> connectionList = connectionMap.getOrDefault(location, new ArrayList<>());
+
+        // remove connections for which the underlying network connection is disconnected.
+        List<Connection> prunedConnectionList = connectionList.stream().filter(connection -> {
+            // Filter out Connection objects which have been completed exceptionally or have been disconnected.
+            CompletableFuture<SessionHandler> r = connection.getSessionHandler();
+            return !r.isDone() || Futures.isSuccessful(r) && r.join().isConnectionEstablished();
+        }).collect(Collectors.toList());
+        log.debug("List of connections to {} that can be used: {}", location, prunedConnectionList);
+
+        // Fetch the Connection related stats.
+        ConnectionSummaryStats stats = prunedConnectionList.parallelStream().collect(collectorStats);
+        // Choose the connection with the least number of sessions.
+        Optional<Connection> suggestedConnection = stats.getConnectionWithMinimumSession(location);
+
+        final Connection connection;
+        if (suggestedConnection.isPresent() && stats.getConnectionCount(location) == clientConfig.getMaxConnectionsPerSegmentStore()) {
+            log.debug("Reusing connection: {}", suggestedConnection.get());
+            Connection oldConnection = suggestedConnection.get();
+            prunedConnectionList.remove(oldConnection);
+            connection = new Connection(oldConnection.getUri(), oldConnection.getSessionHandler(), oldConnection.getSessionCount() + 1);
+        } else {
+            // create a new connection.
+            log.debug("Creating a new connection to {}", location);
+            CompletableFuture<SessionHandler> sessionHandlerFuture = establishConnection(location);
+            connection = new Connection(location, sessionHandlerFuture, 1);
+        }
+        prunedConnectionList.add(connection);
+        connectionMap.put(location, prunedConnectionList);
+        return connection.getSessionHandler().thenApply(sessionHandler -> sessionHandler.createSession(session, rp));
+    }
+
+    @Override
+    public CompletableFuture<ClientConnection> getClientConnection(PravegaNodeUri location, ReplyProcessor rp) {
+        Preconditions.checkNotNull(location, "Location");
+        Preconditions.checkNotNull(rp, "ReplyProcessor");
+        Exceptions.checkNotClosed(closed.get(), this);
+
+        // create a new connection.
+        CompletableFuture<SessionHandler> sessionHandlerFuture = establishConnection(location);
+        Connection connection = new Connection(location, sessionHandlerFuture, 1);
+        return connection.getSessionHandler().thenApply(sessionHandler -> sessionHandler.createConnectionWithSessionDisabled(rp));
+    }
+
+    @Override
+    public int getActiveChannelCount() {
+        return (int) this.channelGroup.stream()
```

Comment:
Why do you `log.debug` every time you invoke `getActiveChannelCount`?

An external user would expect this to return a number and not have a side effect.

Also, I don't know how big this `channelGroup` is; you might consider keeping a private field where you store this count instead of counting every time, which could get expensive quickly.

Reply:
ChannelGroup is a thread-safe Set that contains open Channels and provides different bulk operations on them. The life cycle of the added Channel is handled by the ChannelGroup implementation.
Updated the code to remove this code duplication.